crates/ide-assists/src/handlers/replace_named_generic_with_impl.rs

```diff
@@ -0,0 +1,213 @@
+use hir::Semantics;
+use ide_db::{
+    base_db::{FileId, FileRange},
+    defs::Definition,
+    search::SearchScope,
+    RootDatabase,
+};
+use syntax::{
+    ast::{self, make::impl_trait_type, HasGenericParams, HasName, HasTypeBounds},
+    ted, AstNode,
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: replace_named_generic_with_impl
+//
+// Replaces named generic with an `impl Trait` in function argument.
+//
+// ```
+// fn new<P$0: AsRef<Path>>(location: P) -> Self {}
+// ```
+// ->
+// ```
+// fn new(location: impl AsRef<Path>) -> Self {}
+// ```
+pub(crate) fn replace_named_generic_with_impl(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    // finds `<P: AsRef<Path>>`
+    let type_param = ctx.find_node_at_offset::<ast::TypeParam>()?;
+    // returns `P`
+    let type_param_name = type_param.name()?;
+
+    // The list of type bounds / traits: `AsRef<Path>`
+    let type_bound_list = type_param.type_bound_list()?;
+
+    let fn_ = type_param.syntax().ancestors().find_map(ast::Fn::cast)?;
+    let params = fn_
+        .param_list()?
+        .params()
+        .filter_map(|param| {
+            // function parameter type needs to match generic type name
+            if let ast::Type::PathType(path_type) = param.ty()? {
+                let left = path_type.path()?.segment()?.name_ref()?.ident_token()?.to_string();
+                let right = type_param_name.to_string();
+                if left == right {
+                    Some(param)
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
```

Comment:
If I'm not mistaken this fails to consider params whose type contains `P` in their generic arguments, like:

```rust
fn foo<P: Trait>(
    _: P, // this will be replaced
    _: Option<P>, // this and following `P` would dangle after this assist
    _: impl Iterator<Item = P>,
    _: &dyn Iterator<Item = P>,
) {}
```

Either also collect them and replace the generic params in them, or we can just skip doing this (by not providing this assist if there are such params) for this initial implementation.

If I've missed things and they are actually supported, could you add some tests?

Reply:
Thanks for the list, this is really helpful. I hope to find a way to evaluate them so they are covered too.